### PR TITLE
SDKS 3618- New React Native Project

### PR DESCRIPTION
**Pull request by jeyanthanperiyasamy**

### CODE REVIEW

File: reactnative/reactnative-todo/android/app/src/main/java/com/reactnativetodo/FRAuthSampleBridge.kt [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+    var listener: NodeListener<*>? = null
+    var reactNativePromise: Promise? = null
+
+    override fun getName(): String {
+        return "FRAuthSampleBridge"
+    }
+
+
+    @ReactMethod
+    fun start(promise: Promise) {
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/android/app/src/main/java/com/reactnativetodo/FRAuthSampleBridge.kt:69]: i agree, we need that, i like to do in a different story

File: reactnative/reactnative-todo/src/components/journey/journey-handler.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * journey-hook.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/journey/journey-handler.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/journey/journey-handler.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+
+import { FRStep } from '@forgerock/javascript-sdk';
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/journey/journey-handler.js:11]: Why are we using JS SDK here?

File: reactnative/reactnative-todo/src/components/journey/utilities.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * failed-policies.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/journey/utilities.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/todo/add-todo.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * add-todo.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/todo/add-todo.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/todo/edit-todo.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * edit-todo.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/todo/edit-todo.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/todo/reducer.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * reducer.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/todo/reducer.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/todo/todo.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * todo.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/todo/todo.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/todo/toggle.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * toggle.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/todo/toggle.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/utilities/alert.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * alert.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/utilities/alert.js:6]: Update copyright

File: reactnative/reactnative-todo/src/components/utilities/loading.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * loading.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/components/utilities/loading.js:6]: Update copyright

File: reactnative/reactnative-todo/src/screens/logout.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
+/*
+ * forgerock-react-native-sample
+ *
+ * logout.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/src/screens/logout.js:6]: Update copyright

**Review by rodrigoareis — CHANGES_REQUESTED** [review]
Overall changes looks good, left some comments to address

**Review by george-bafaloukas-forgerock — COMMENTED** [review]
@jeyanthanperiyasamy ping me to discuss the comments when free. We will need on the back of that to create an update for the RN Tutorial for docs for Chris as well

File: reactnative/README.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+3. **Clone the Repository**:
+    ```bash
+    git clone https://github.com/ForgeRock/sdk-sample-apps.git
+    ```
+
+4. **Open the React Native Project** in Visual Studio Code.
+
+5. **Configure iOS**:
+    - Open the `FRAuthSampleBridge.swift` file.
+    - Locate `FRAuth.plist` and replace placeholder strings with your **OAuth 2.0 application** details.
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/README.md:51]: https://pingidentity.atlassian.net/browse/SDKS-3696 - we have a seperate story to deal dynamic config. its out of scope for the existing story

File: reactnative/README.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+6. **Configure Android**:
+    - Open the `FRAuthSampleBridge.kt` file.
+    - Locate `Strings.xml` and update placeholder strings with your **OAuth 2.0 application** details.
+
+7. **Set API Base URL**:
+    - Edit `.env.js` and set `API_BASE_URL` to the IP and port of your `todo-api` backend:
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/README.md:58]: i rebase the main branch..is it not up do date ?

File: reactnative/README.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+### Configuration Variables for Android/iOS
+
+Update the following environment variables in `strings.xml` for Android and `Info.plist` for iOS dynamically:
+
+| Variable                         | Description                                    | Example Value                                      |
+|-----------------------------------|------------------------------------------------|----------------------------------------------------|
+| `PINGAM_OAUTH_CLIENT_ID`          | OAuth 2.0 Client ID                            | `AndroidTest` / `iOSClient`                        |
+| `PINGAM_OAUTH_REDIRECT_URI`       | OAuth 2.0 Redirect URI                         | `org.ping.demo:/oauth2redirect` / `myapp://oauth2redirect` |
+| `PINGAM_URL`                      | PingAM Instance URL                            | `https://openam-xvc.forgeblock.com/am`             |
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/README.md:76]: openam-xvc.forgeblock.com/am is a dummy url

File: reactnative/reactnative-todo/ios/reactnativetodo/FRAuthConfig.plist [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+	<key>forgerock_oauth_redirect_uri</key>
+	<string>http://www.example.com</string>
+	<key>forgerock_oauth_scope</key>
+	<string>openid profile email address</string>
+	<key>forgerock_oauth_url</key>
+	<string>https://idc.petrov.ca/am</string>
+	<key>forgerock_oauth_threshold</key>
+	<string>60</string>
+	<key>forgerock_url</key>
+	<string>https://openam-sdks.forgeblocks.com/am</string>
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/ios/reactnativetodo/FRAuthConfig.plist]: i forgot to update this

File: reactnative/reactnative-todo/src/components/journey/journey-handler.js (hunk 3 above)
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/src/components/journey/journey-handler.js:11]: ha good question we use FRStep from javascript-sdk , which has some parsing logic..i have to look this deep and see if we can remove..but again not in this story

import { FRStep } from '@forgerock/javascript-sdk';

 const step = new FRStep(next);

File: reactnative/reactnative-todo/android/app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -0,0 +1,33 @@
+<resources>
+   
+ <string name="app_name">reactnativetodo</string>
+  <!-- OAuth 2.0 Client Details -->
+ <string name="forgerock_oauth_client_id" translatable="false">${PINGAM_OAUTH_CLIENT_ID}</string>
+ <string name="forgerock_oauth_redirect_uri" translatable="false">${PINGAM_OAUTH_REDIRECT_URI}</string>
+<string name="forgerock_oauth_scope" translatable="false">openid profile email address phone</string>
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/android/app/src/main/res/values/strings.xml]: moved to dynamic config in configuration.kt

File: reactnative/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/README.md:51]: moved to dynamic config in configuration.kt and configuration.swift

File: reactnative/README.md [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+    ```
+
+4. **Open the React Native Project** in Visual Studio Code.
+
+5. **Configure iOS**:
+    - Open the `FRAuthSampleBridge.swift` file.
+    - Locate `FRAuth.plist` and replace placeholder strings with your **OAuth 2.0 application** details.
+
+6. **Configure Android**:
+    - Open the `FRAuthSampleBridge.kt` file.
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/README.md:54]: moved to dynamic config in configuration.kt and configuration.swift

File: reactnative/reactnative-todo/ios/reactnativetodo/FRAuthConfig.plist (hunk 16 above)
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/ios/reactnativetodo/FRAuthConfig.plist]: removed the plist now and moved to dynamic config in configuration.swift

File: reactnative/reactnative-todo/.env.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Avoid trailing slashes in the URL string values below
+ */
+const AM_URL = 'https://openam-sdks.forgeblocks.com/am';
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/.env.js:4]: we can cleanup this file, since we are moved the todo in javascript folder

**Comment by jeyanthanperiyasamy** [comment]
Addressed most  comments. please verify @rodrigoareis @george-bafaloukas-forgerock

File: reactnative/reactnative-todo/.env.js [hunk 20]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Avoid trailing slashes in the URL string values below
+ */
+const API_PORT = 9443;
+const API_BASE_URL = 'http://10.0.2.2'; // for iOS - http://localhost
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/.env.js:5]: Do we need this ip address and port here? Can we replace with placeholder?

File: reactnative/reactnative-todo/android/app/src/main/java/com/reactnativetodo/Configuration.kt [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+
+package com.reactnativetodo
+
+
+/**
+ * Configuration class for the ForgeRock SDK.
+ */
+data class Configuration(
+    /** The main authentication journey name.  */
+    val mainAuthenticationJourney: String = "jey-webAuthn",
```
Comment (rodrigoareis) [inline reactnative/reactnative-todo/android/app/src/main/java/com/reactnativetodo/Configuration.kt:16]: Can we replace this config with placeholder? We should expose details of our environments publicly

Comment (george-bafaloukas-forgerock) [reply reactnative/reactnative-todo/android/app/src/main/java/com/reactnativetodo/Configuration.kt:16]: I agree replace all values with [Placeholders]

Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/android/app/src/main/java/com/reactnativetodo/Configuration.kt]: ha i accidentally committed that ..good catch . Fixed now

File: reactnative/reactnative-todo/.env.js (hunk 20 above)
Comment (jeyanthanperiyasamy) [reply reactnative/reactnative-todo/.env.js]: fixed as a placeholder.

**Review by george-bafaloukas-forgerock — APPROVED** [review]
LGTM

**Review by rodrigoareis — APPROVED** [review]
Changes looks good to me

**Pull request merged**